homeassistant/components/huawei_lte/device_tracker.py

```diff
@@ -2,34 +2,95 @@
+    DOMAIN as DEVICE_TRACKER_DOMAIN,
+    SOURCE_TYPE_ROUTER,
+)
 from homeassistant.components.device_tracker.config_entry import ScannerEntity
 from homeassistant.const import CONF_URL
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
 from . import HuaweiLteBaseEntity
-from .const import DOMAIN, KEY_WLAN_HOST_LIST
+from .const import DOMAIN, KEY_WLAN_HOST_LIST, UPDATE_SIGNAL
 
 
 _LOGGER = logging.getLogger(__name__)
 
+_NEW_DEVICE_SCAN = "new_device_scan"
+
 
 async def async_setup_entry(hass, config_entry, async_add_entities):
     """Set up from config entry."""
+
+    # Grab hosts list once to examine whether the initial fetch has got some data for
+    # us, i.e. if wlan host list is supported. Only set up a subscription and proceed
+    # with adding and tracking entities if it is.
     router = hass.data[DOMAIN].routers[config_entry.data[CONF_URL]]
+    try:
+        _ = router.data[KEY_WLAN_HOST_LIST]["Hosts"]["Host"]
+    except KeyError:
+        _LOGGER.debug("%s[%s][%s] not in data", KEY_WLAN_HOST_LIST, "Hosts", "Host")
+        return
+
+    # Set of entities we've added already
+    tracked: Set[str] = set()
+
+    # Tell parent router to grab hosts list so we can get new entities
+    router.subscriptions[KEY_WLAN_HOST_LIST].add(
+        f"{DEVICE_TRACKER_DOMAIN}/{_NEW_DEVICE_SCAN}"
+    )
+
+    async def _async_maybe_add_new_entities(url: str) -> None:
+        """Add new entities if the update signal comes from our router."""
+        if url == router.url:
+            await async_add_new_entities(hass, url, async_add_entities, tracked)
+
+    # Register to handle router data updates
+    disconnect_dispatcher = async_dispatcher_connect(
+        hass, UPDATE_SIGNAL, _async_maybe_add_new_entities
+    )
+    router.signal_handlers[
+        f"{DEVICE_TRACKER_DOMAIN}_disconnect"
+    ] = disconnect_dispatcher
+
+    # Add new entities waiting to be added from initial scan
+    await async_add_new_entities(hass, router.url, async_add_entities, tracked)
+
+
+async def async_add_new_entities(hass, router_url, async_add_entities, tracked):
+    """Add new entities."""
+    router = hass.data[DOMAIN].routers[router_url]
     try:
         hosts = router.data[KEY_WLAN_HOST_LIST]["Hosts"]["Host"]
-    except (KeyError, TypeError):
+    except KeyError:
         _LOGGER.debug("%s[%s][%s] not in data", KEY_WLAN_HOST_LIST, "Hosts", "Host")
         return
 
-    entities = []
+    new_entities = []
     for host in (x for x in hosts if x.get("MacAddress")):
-        entities.append(HuaweiLteScannerEntity(router, host["MacAddress"]))
-    async_add_entities(entities, True)
+        entity = HuaweiLteScannerEntity(router, host["MacAddress"])
+        if entity.unique_id in tracked:
+            continue
+        tracked.add(entity.unique_id)
+        new_entities.append(entity)
+    async_add_entities(new_entities, True)
+
+
+async def async_unload_entry(hass, config_entry):
```

Comment:
Platforms don't support `async_unload_entry`. We have to unsubscribe from `async_unload_entry` in the component.

Reply:
Oh, that's kind of surprising given that unloading needs to be forwarded to platforms. But handled in component now.